### PR TITLE
Fix extended functional tests

### PR DESCRIPTION
**Pull request by Naviabheeman**
Revert back changing mutex to recursivemutex in net.cpp as it caused deadlock in maxupload limit test.
Enable executing all extended functional tests from CI
Fix all failing extended functional tests

### CODE REVIEW

File: test/functional/feature_maxuploadtarget.py [hunk 1]
```diff
@@ -33,48 +37,97 @@ def on_block(self, message):
...
         self.nodes[0].generate(31, self.signblockprivkey_wif)
+        self.utxo_cache = self.nodes[0].listunspent()
+        self.sync_all()
 
-        # p2p_conns[0] will only request old blocks
-        # p2p_conns[1] will only request new blocks
-        # p2p_conns[2] will test resetting the counters
+        # p2p_conns[0] will only request old blocks and resetting the counters
+        # p2p_conns[1] will request new and old blocks
+        # p2p_conns[2] will test whitelist
```
Comment (azuchi) [inline test/functional/feature_maxuploadtarget.py:97]: Why these change necessary?

File: test/functional/feature_maxuploadtarget.py [hunk 2]
```diff
@@ -33,48 +37,97 @@ def on_block(self, message):
...
 
+        self.nodes[0].generate(100, self.signblockprivkey_wif)
+        self.sync_all()
+
         # Advance to two days ago
-        self.nodes[0].setmocktime(int(time.time()) - 2*60*60*24)
+        self.mocktime =  int(time.time()) - 2*60*60*24
+        self.nodes[0].setmocktime(self.mocktime)
+        self.nodes[1].setmocktime(self.mocktime)
+        self.nodes[2].setmocktime(self.mocktime)
```
Comment (azuchi) [inline test/functional/feature_maxuploadtarget.py:119]: Bitcoin also does not seem to mock the time of nodes 1 and 2, but what is the intention of this change?

File: test/functional/feature_pruning.py [hunk 3]
```diff
@@ -34,15 +68,15 @@ def set_test_params(self):
 
         # Create nodes 0 and 1 to mine.
         # Create node 2 to test pruning.
-        self.full_node_default_args = ["-maxreceivebuffer=20000", "-checkblocks=5", "-limitdescendantcount=100", "-limitdescendantsize=5000", "-limitancestorcount=100", "-limitancestorsize=5000" ]
+        self.full_node_default_args = ["-dev", "-maxreceivebuffer=20000", "-checkblocks=5", "-limitdescendantcount=100", "-limitdescendantsize=5000", "-limitancestorcount=100", "-limitancestorsize=5000" ]
```
Comment (azuchi) [inline test/functional/feature_pruning.py:71]: Why `-dev` option necessary?

**Review by azuchi — CHANGES_REQUESTED** [review]

File: test/functional/test_runner.py [hunk 4]
```diff
@@ -188,9 +188,10 @@
 ]
 
 EXTENDED_SCRIPTS = [
+    # disabling as this script does not invoke the intended test path
+    #'feature_pruning.py',
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply test/functional/test_runner.py:192]: I added the reason as a comment in the test case:
```
Tapyrus  update:
This test was designed without compact blocks.  now all block announcements on tapyrus happen using compact blocks. So the steps in this test which are designed to trigger pruning are not causing the expected behaviour.  It is not possible to disable compact blocks with parameters.
```

File: test/functional/test_runner.py [hunk 5]
```diff
@@ -205,7 +206,6 @@
     'feature_bip68_sequence.py --scheme SCHNORR',
     'mining_getblocktemplate_longpoll.py',
     'p2p_timeouts.py',
-    'feature_blockheaderxfield.py',
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply test/functional/test_runner.py:208]: This test was added as test for serialisation and deserialisation of blocks with known and unknown xfields. After the addition of `max-blocksize` we have more detailed tests. I intended to delete the test case as it is not needed anymore. `feature_federation_management.py` and `feature_xfield_maxblocksize.py` are a lot more detailed.

File: test/functional/test_framework/util.py [hunk 6]
```diff
@@ -534,7 +534,7 @@ def gen_return_txouts():
     # Some pre-processing to create a bunch of OP_RETURN txouts to insert into transactions we create
     # So we have big transactions (and therefore can't fit very many into each block)
     # create one script_pubkey
-    script_pubkey = "6a4d0200"  # OP_RETURN OP_PUSH2 512 bytes
+    script_pubkey = "754d0200"  # OP_DROP OP_PUSH2 512 bytes
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply test/functional/test_framework/util.py:537]: It was probably a change I missed to cleanup. It was not intentional

File: test/functional/wallet_txn_clone.py [hunk 7]
```diff
@@ -143,7 +146,7 @@ def run_test(self):
         # Check node0's total balance; should be same as before the clone
         # less possible orphaned matured subsidy
         if (self.options.mine_block):
-            expected -= 50
+            expected -= coinbase['amount']
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply test/functional/wallet_txn_clone.py:149]: This test is different in bitcoin and tapyrus. In bitcoin there is a `segwit` mode where it tests that the node does not allow transaction malleability and stops there. Node balances are not verified. Our test is different after this step - all results are different. 
so I can only guess that this test was never executed with `--mine-block` option in tapyrus before.

File: test/functional/feature_assumevalid.py [hunk 8]
```diff
@@ -13,25 +11,24 @@
...
     0:        genesis block
     1:        block 1 with coinbase transaction output.
-    2-101:    bury that block with 100 blocks so the coinbase transaction
-              output can be spent
-    102:      a block containing a transaction spending the coinbase
+    2:      a block containing a transaction spending the coinbase
               transaction output. The transaction has an invalid signature.
-    103-2202: bury the bad block with just over two weeks' worth of blocks
-              (2100 blocks)
+    3-6102: bury the bad block with 6100 blocks:
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply test/functional/feature_assumevalid.py:16]: I think I was able to meet the hardcoded limit, `60 * 60 * 24 * 7 * 2 more than block2` with only those many blocks. ```/src/validation.cpp:1929```

File: test/functional/feature_bip68_sequence.py [hunk 9]
```diff
@@ -31,70 +31,17 @@ def run_test(self):
         # Generate some coins
         self.nodes[0].generate(11, self.signblockprivkey_wif)
 
-        self.log.info("Running test disable flag")
-        self.test_disable_flag()
-
         self.log.info("Running test sequence-lock-confirmed-inputs")
-        self.test_sequence_lock_confirmed_inputs()
+        #self.test_sequence_lock_confirmed_inputs()
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply test/functional/feature_bip68_sequence.py:35]: I think it's another shortcut for testing which was added unintentionally. will fix it.

File: test/functional/feature_fee_estimation.py [hunk 10]
```diff
@@ -72,7 +72,7 @@ def small_txpuzzle_randfee(from_node, conflist, unconflist, amount, min_fee, fee
 
     return (ToHex(tx), fee)
 
-def split_inputs(from_node, txins, txouts, initial_split=False):
+def split_inputs(from_node, txins, txouts, scheme, initial_split=False):
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply test/functional/feature_fee_estimation.py:75]: Actually all our functional tests run with ECDSA and with SCHNORR signature independently. we use them in the signing rpcs. In this test `self` is not even defined where the change was made initially. again it was not caught as this test was never executed.

File: test/functional/feature_maxuploadtarget.py (hunk 2 above)
Comment (Naviabheeman) [reply test/functional/feature_maxuploadtarget.py:119]: When we use generate rpc, multiple blocks with the same timestamp skews the clock and later blocks are rejected because of their timestamp. mock was needed to. synchronise. the nodes too.

File: test/functional/feature_maxuploadtarget.py (hunk 1 above)
Comment (Naviabheeman) [reply test/functional/feature_maxuploadtarget.py:97]: When using peer2 initially, it did not restart.  I guess the problem was timestamp related and needed -reindex. So I just used peer0 to very the resetting of counters. Once the test was passing I didn't backtrack this change.

File: test/functional/feature_pruning.py (hunk 3 above)
Comment (Naviabheeman) [reply test/functional/feature_pruning.py:71]: No it's not necessary, all functional test nodes run in 'dev' mode. It's another attempt to run this test.  none of these changes make any difference to this test.

File: test/functional/feature_rbf.py [hunk 11]
```diff
@@ -71,14 +71,16 @@ def set_test_params(self):
                            "-limitancestorcount=50",
                            "-limitancestorsize=101",
                            "-limitdescendantcount=200",
-                           "-limitdescendantsize=101"],
-                           ["-mempoolreplacement=0"]]
+                           "-limitdescendantsize=101",
+                           "-txindex"],
+                           ["-mempoolreplacement=0",
+                            "-txindex"]]
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply test/functional/feature_rbf.py:77]: You are right. is not necessary!

File: test/functional/feature_fee_estimation.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (azuchi) [reply test/functional/feature_fee_estimation.py:75]: yes. However, it need to test both in the signing rpc functional test only. The test will be over soon. Well, let's organize it with another PR.

File: test/functional/test_runner.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (azuchi) [reply test/functional/test_runner.py:192]: Why can't test pruning when using compact blocks? You can use the pruning feature even if you use compact block, right?

File: test/functional/wallet_txn_clone.py [hunk 12]
```diff
@@ -90,10 +90,13 @@ def run_test(self):
...
         # matured block, minus tx1 and tx2 amounts, and minus transaction fees:
         expected = starting_balance + node0_tx1["fee"] + node0_tx2["fee"]
+        coinbase = {}
         if self.options.mine_block:
-            expected += 50
-        expected += tx1["amount"] + tx1["fee"]
-        expected += tx2["amount"] + tx2["fee"]
+            block = self.nodes[0].getblock(self.nodes[0].getbestblockhash())
+            coinbase = self.nodes[0].gettransaction(block['tx'][0])
+            expected += coinbase['amount']
```
Comment (azuchi) [inline test/functional/wallet_txn_clone.py:97]:
```py
# Node0's balance should be starting balance, plus 50TPC for another
# matured block, minus tx1 and tx2 amounts, and minus transaction fees:
```

If you test according to this test scenario, don't you need Coinbase transactions and can leave them as they are?

File: test/functional/wallet_txn_clone.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (azuchi) [reply test/functional/wallet_txn_clone.py:149]: I mean, is it okay to leave it at 50 as in the comment above?

File: test/functional/feature_assumevalid.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (azuchi) [reply test/functional/feature_assumevalid.py:16]: But why hasn't Bitcoin made this change?

File: test/functional/feature_maxuploadtarget.py (hunk 2 above)
Comment (azuchi) [reply test/functional/feature_maxuploadtarget.py:119]: Why hasn't it changed in Bitcoin?

**Review by azuchi — CHANGES_REQUESTED** [review]

File: test/functional/test_framework/util.py [hunk 13]
```diff
@@ -558,7 +558,7 @@ def create_lots_of_big_transactions(node, txouts, utxos, num, fee):
         t = utxos.pop()
         inputs = [{"txid": t["txid"], "vout": t["vout"]}]
         outputs = {}
-        change = t['amount'] - fee
+        change = t['amount'] - fee - 10
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply test/functional/test_framework/util.py:561]: I think this too was a consequence of the change in the previous command. I'm reverting this too.

File: test/functional/wallet_txn_clone.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply test/functional/wallet_txn_clone.py:149]: No. In tapyrus the when the block is mined the block reward is more than 50 because of one transaction other than coinbase.

File: test/functional/feature_assumevalid.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply test/functional/feature_assumevalid.py:16]: We have seen similar problems in the past - #248. Our time gets skewed when blocks are mined continuously. Bitcoin does not seem to have such issue. And I noticed that `GetAdjustedTime()` seems to have been removed from its code recently. Maybe we can check if that can fix our skew.

File: test/functional/feature_maxuploadtarget.py (hunk 2 above)
Comment (Naviabheeman) [reply test/functional/feature_maxuploadtarget.py:119]: The clock skew again is tapyrus specific as mentioned in the other comment: https://github.com/chaintope/tapyrus-core/pull/297#discussion_r1541380127

File: test/functional/wallet_txn_clone.py (hunk 12 above)
Comment (Naviabheeman) [reply test/functional/wallet_txn_clone.py:97]: sorry. I missed this as next comment is also related to the same node balance. In our test case while mining a block the non maleated transactions are confirmed and add fee also to the coinbase amount. I'll update the comments.

Comment (azuchi) [reply test/functional/wallet_txn_clone.py:97]: Are the fees different from those added in the code below?

```
expected += (tx1["amount"] + tx1["fee"])
expected += (tx2["amount"] + tx2["fee"])
```

File: test/functional/wallet_txn_clone.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (azuchi) [reply test/functional/wallet_txn_clone.py:149]: yes. Coinbase includes block reward (50BTC) and fees. In the original test case, these are clearly separated, so wouldn't that make the test more meaningful?

File: test/functional/wallet_txn_clone.py (hunk 12 above)
Comment (Naviabheeman) [reply test/functional/wallet_txn_clone.py:97]: This test runs in two modes. When 'mine_block' is enabled, tx1 and tx2 get mined and confirmed. In the default mode these two remain in the mempool. This is why there is a difference in the balances. In bitcoin, mine_block mode does not execute the test till the end. It just verifies the clones not mined midway and exits.

File: test/functional/wallet_txn_clone.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply test/functional/wallet_txn_clone.py:149]: ok, I think I get what you mean only now! When in mine_block mode, the wallet balance is includes the confirmed block.  The difference is in the number of utxos.

**without mine block:**

| wallet balance | utxos |
|--|--|
| initial - tx1_amount - tx1_fee  - tx2_amount - tx2_fee | initial + 4 unconfirmed |

**mine block:**

| wallet balance | utxos |
|--|--|
| initial + 50 (tx1 and tx2 balances are added the same node) | initial + 4 confirmed |

File: test/functional/wallet_txn_clone.py (hunk 12 above)
Comment (Naviabheeman) [reply test/functional/wallet_txn_clone.py]: I updated the calculations again. Hope its cleaner and clearer now.

File: test/functional/wallet_txn_clone.py [hunk 14]
```diff
@@ -11,7 +11,10 @@
     connect_nodes,
     disconnect_nodes,
     sync_blocks,
+    hex_str_to_bytes
 )
+from test_framework.messages import CTransaction
+from io import BytesIO
```
Comment (azuchi) [inline test/functional/wallet_txn_clone.py:17]: These don't seem to be used, so there's no need to add them.

**Review by azuchi — CHANGES_REQUESTED** [review]

**Review by azuchi — APPROVED** [review]
